fs/fs.go

```diff
@@ -74,20 +74,79 @@ import (
 	"github.com/opencontainers/go-digest"
 	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 	"github.com/sirupsen/logrus"
+	"golang.org/x/sync/semaphore"
 )
 
 var (
 	defaultIndexSelectionPolicy = SelectFirstPolicy
 	fusermountBin               = "fusermount"
+	preresolverQueueBufferSize  = 1024 // arbitrarily chosen buffer size
 )
 
+// Preresolver will resolve a number of layers in parallel,
+// up to the amount specified by MaxConcurrency.
+type preresolver struct {
+	queue chan func(context.Context) string
+	cache *sync.Map
+	smp   *semaphore.Weighted
+}
+
+func newPreresolver(maxConcurrency int64) *preresolver {
+	pr := &preresolver{}
+	pr.queue = make(chan func(context.Context) string, preresolverQueueBufferSize)
+	pr.cache = &sync.Map{}
+	if maxConcurrency > 0 {
+		pr.smp = semaphore.NewWeighted(maxConcurrency)
+	}
+	return pr
+}
+
+func (pr *preresolver) Start(ctx context.Context) error {
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				resolveFn := <-pr.queue
+				if pr.smp != nil {
```

Comment:
Can you pls add a comment at this line something like
`when the concurrency limit is enabled then wait for the turn`